stl/inc/syncstream

```diff
@@ -0,0 +1,329 @@
+#pragma push_macro("new")
+#pragma push_macro("emit")
+#undef new
+#undef emit
+
+_STD_BEGIN
+
+extern "C" shared_mutex& __cdecl _Get_mutex_for_instance(void* _Ptr);
+extern "C" void __cdecl _Acquire_mutex_for_instance(void* _Ptr) noexcept;
+extern "C" void __cdecl _Release_mutex_for_instance(void* _Ptr) noexcept;
+
+// CLASS TEMPLATE basic_syncbuf
+template <class _Elem, class _Traits, class _Alloc>
+class basic_syncbuf : public basic_streambuf<_Elem, _Traits> {
+public:
+    using int_type       = typename _Traits::int_type;
+    using pos_type       = typename _Traits::pos_type;
+    using off_type       = typename _Traits::off_type;
+    using allocator_type = _Alloc;
+    using streambuf_type = basic_streambuf<_Elem, _Traits>;
+
+    using pointer   = typename allocator_traits<_Alloc>::pointer;
+    using size_type = typename allocator_traits<_Alloc>::size_type;
+
+    basic_syncbuf() = default;
+
+    explicit basic_syncbuf(streambuf_type* _Strbuf) : basic_syncbuf(_Strbuf, _Alloc{}) {}
+
+    basic_syncbuf(streambuf_type* _Strbuf, const _Alloc& _Al_) : _Wrapped(_Strbuf), _Al(_Al_) {
+        if (_Wrapped != nullptr) {
+            _Acquire_mutex_for_instance(_Wrapped);
+        }
+        _Init();
+    }
+
+    basic_syncbuf(basic_syncbuf&& _Right) {
+        _Al = _STD move(_Right._Al);
+        _Assign_rv(_STD move(_Right));
+    }
+
+    ~basic_syncbuf() {
+        _Emit();
+        _Tidy();
+    }
+
+    basic_syncbuf& operator=(basic_syncbuf&& _Right) {
+        _Emit();
+        _Move_assign(_STD move(_Right), _Choose_pocma<_Alloc>{});
+        return *this;
+    }
+
+    void swap(basic_syncbuf& _Right) noexcept {
+        if (this != _STD addressof(_Right)) {
+            _Pocs(_Getal(), _Right._Getal());
+            _Swap_except_al(_Right);
+        }
+    }
+
+    bool emit() {
+        if (_Wrapped == nullptr) {
+            return false;
+        }
+
+        bool _Result                 = true;
+        const size_type _Data_size   = _Get_data_size();
+        const pointer _Begin_seq_ptr = pointer_traits<pointer>::pointer_to(*streambuf_type::pbase());
+        if (_Data_size > 0 || _Record_sync) {
+            scoped_lock _Guard(_Get_mutex_for_instance(_Wrapped));
+
+            if (_Data_size > 0 && _Data_size != _Wrapped->sputn(_Begin_seq_ptr, _Data_size)) {
+                _Result = false;
+            }
+
+            if (_Record_sync) {
+                if (_Wrapped->pubsync() == -1) {
+                    _Result = false;
+                }
+            }
+        }
+        _Record_sync = false;
+        streambuf_type::setp(
+            _Begin_seq_ptr, pointer_traits<pointer>::pointer_to(*streambuf_type::epptr())); // reset written data
+        return _Result;
+    }
+
+    _NODISCARD streambuf_type* get_wrapped() const noexcept {
+        return _Wrapped;
+    }
+
+    _NODISCARD allocator_type get_allocator() const noexcept {
+        return _Getal();
+    }
+
+    void set_emit_on_sync(bool _Val) noexcept {

```

Comment:
`set_emit_on_sync` is modifying a member variable, it should not be right to mark as const 👀.

Reply:
I think he meant to make `_Val` const.

Reply:
Again, I misunderstood the comment, my bad 🤣. It could be `const`, but this will result in a different signature than [P0053R7](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0053r7.pdf). Is it allowed to do such changes?